File: packages/types/src/host/host-bridge.ts
/**
 * host/host-bridge.ts
 *
 * The canonical, type-only renderer→host contract. One interface, multiple
 * adapters (Tauri now; Electron in Plan 2). Events are subscription functions
 * returning an Unsubscribe so the transport (Tauri Channel/listen vs IPC) stays
 * inside the adapter.
 *
 * Plan 1 scope: app / fs / shell / notify / terminal / preview / daemon / log.
 * Preview is the `mount()` seam (Plan 2).
 * Plan 3 scope: updates + presence — all three adapters now implement both.
 */

export type Unsubscribe = () => void;

import type { Platform, DaemonStatus, LogLevel, UpdateStatus, PresenceState } from './host-contract.js';
import type { AppInfoSchema, RegionSchema } from './host-contract.js';
import type { z } from 'zod';
export type { Platform, DaemonStatus, LogLevel, UpdateStatus, PresenceState } from './host-contract.js';

export type AppInfo = z.infer<typeof AppInfoSchema>;

export interface Bounds {
  x: number;
  y: number;
  w: number;
  h: number;
}

export type Region = z.infer<typeof RegionSchema>;

export interface InspectResult {
  tabId: string;
  selector: string | null;
  rect: Bounds | null;
  viewport: Bounds | null;
}

export interface RegionSelectResult {
  tabId: string;
  /** Selected region in webview-viewport CSS px, or null when cancelled (Escape / zero-size). */
  region: Region | null;
}

export interface TerminalOpts {
  id: string;
  cwd: string;
  cols: number;
  rows: number;
}

export interface TerminalHandlers {
  onData: (bytes: Uint8Array) => void;
  onExit: (code: number | null) => void;
}

export interface TerminalHandle {
  write(data: string): Promise<void>;
  resize(cols: number, rows: number): Promise<void>;
  kill(): Promise<void>;
}

export interface PreviewOpts {
  /** Selects the persistent session partition: persist:sandbox-{projectId} (Electron). */
  projectId?: string;
  /** Initial frame; the renderer toggles it via handle.setDevice. */
  device?: 'desktop' | 'mobile';
}

/**
 * A mounted preview surface. The renderer reserves a DOM container and hands it to
 * mount(); the handle owns one backing webview.
 *
 * Coordinate space: capture() region is CSS pixels in the WEBVIEW VIEWPORT space
 * (top-left = page content origin). The backend (Tauri WKWebView snapshot / Electron
 * capturePage) applies device-pixel-ratio scaling — the renderer never multiplies by DPR.
 *
 * Visibility/occlusion: Tauri composites the webview ABOVE the DOM, so setVisible(false)
 * blanks the OS layer when a DOM overlay covers the region. Electron stacks the <webview>
 * IN the DOM, so setVisible(false) on occlusion is a near-no-op — both hosts must TOLERATE
 * the renderer emitting it (the renderer keeps the existing occlusion logic unchanged).
 */
export interface PreviewHandle {
  setVisible(visible: boolean): void;
  /**
   * True when the backing webview composites ABOVE the DOM (Tauri native child
   * WKWebView) — a DOM overlay over the preview region must hide it. False when
   * the webview is an in-DOM element that respects z-index (Electron <webview>
   * tag) — DOM overlays stack over it and it must NOT be hidden for them.
   */
  readonly compositesAboveDom: boolean;
  navigate(url: string): Promise<void>;
  capture(region?: Region): Promise<Uint8Array>;
  startInspect(): Promise<void>;
  onInspect(cb: (result: InspectResult) => void): Unsubscribe;
  startRegionSelect(): Promise<void>;
  onRegionSelect(cb: (result: RegionSelectResult) => void): Unsubscribe;
  /**
   * Subscribe to navigations that occur inside the preview webview — link
   * clicks, redirects, and SPA route changes. Mirrors the address-bar in a real
   * browser. Returns an Unsubscribe.
   */
  onNavigate(cb: (url: string) => void): Unsubscribe;
  /** Tauri: re-read container.getBoundingClientRect() into the native layer. Electron: no-op. */
  refit(): void;
  setDevice(device: 'desktop' | 'mobile'): void;
  destroy(): void;
}

export interface HostBridge {
  app: {
    getInfo(): Promise<AppInfo>;
    getHomedir(): Promise<string>;
    getAuthToken(): Promise<string | null>;
    platform(): Promise<Platform>;
  };
  fs: {
    readFile(path: string): Promise<string | null>;
    readFileBase64(path: string): Promise<string | null>;
    showItemInFolder(path: string): Promise<void>;
  };
  shell: {
    openExternal(url: string): Promise<void>;
  };
  notify(title: string, body?: string): Promise<void>;
  terminal: {
    create(opts: TerminalOpts, handlers: TerminalHandlers): Promise<TerminalHandle>;
  };
  preview: {
    mount(container: HTMLElement, url: string, opts?: PreviewOpts): PreviewHandle;
    clearSession(projectId: string): Promise<void>;
  };
  daemon: {
    port(): Promise<number>;
    status(): Promise<DaemonStatus>;
    onStatus(cb: (status: DaemonStatus) => void): Promise<Unsubscribe>;
  };
  updates: {
    check(): Promise<UpdateStatus>;
    download(): Promise<void>;
    install(): void;
    onStatus(cb: (s: UpdateStatus) => void): Promise<Unsubscribe>;
  };
  presence: {
    reportActivity(state: PresenceState): Promise<void>;
  };
  log(level: LogLevel, module: string, message: string, data?: unknown): void;
  /** Tauri installs the window-drag listener here; Electron is a CSS no-op. */
  init?(): void;
}

File: packages/core-rs/crates/mainframe-types/src/host/host_bridge.rs
//! Ported from `packages/types/src/host/host-bridge.ts`.
//!
//! The canonical, type-only renderer→host contract. Only the **data** shapes are
//! ported here (Bounds, InspectResult, RegionSelectResult, TerminalOpts,
//! PreviewOpts). The `HostBridge` interface and the handle interfaces
//! (`TerminalHandle`, `PreviewHandle`, `TerminalHandlers`) are async/callback
//! renderer↔host contracts implemented per shell (Tauri/Electron in JS + Rust) and
//! are not daemon-consumed — they are not ported as Rust types.
// TODO(port): HostBridge / TerminalHandle / PreviewHandle / TerminalHandlers are
// behavioral bridge interfaces (Promise-returning methods, DOM handles, callbacks),
// not serde data; they live in the desktop shells, not this crate.

use serde::{Deserialize, Serialize};

pub use super::daemon_target::DaemonMeta;
pub use super::host_contract::{
    AppInfo, DaemonStatus, LogLevel, Platform, PresenceState, Region, UpdateStatus,
};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Bounds {
    pub x: f64,
    pub y: f64,
    pub w: f64,
    pub h: f64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InspectResult {
    pub tab_id: String,
    pub selector: Option<String>,
    pub rect: Option<Bounds>,
    pub viewport: Option<Bounds>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RegionSelectResult {
    pub tab_id: String,
    /// Selected region in webview-viewport CSS px, or null when cancelled.
    pub region: Option<Region>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TerminalOpts {
    pub id: String,
    pub cwd: String,
    pub cols: i64,
    pub rows: i64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum PreviewDevice {
    Desktop,
    Mobile,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct PreviewOpts {
    /// Selects the persistent session partition (Electron).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_id: Option<String>,
    /// Initial frame; the renderer toggles it via handle.setDevice.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub device: Option<PreviewDevice>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn inspect_result_serializes_null_fields() {
        let json = r#"{"tabId":"t1","selector":null,"rect":null,"viewport":null}"#;
        let r: InspectResult = serde_json::from_str(json).unwrap();
        assert!(r.selector.is_none());
        assert_eq!(serde_json::to_string(&r).unwrap(), json);
    }

    #[test]
    fn region_select_result_round_trips() {
        let json = r#"{"tabId":"t1","region":{"x":0.0,"y":0.0,"w":100.0,"h":50.0}}"#;
        let r: RegionSelectResult = serde_json::from_str(json).unwrap();
        assert!(r.region.is_some());
        assert_eq!(serde_json::to_string(&r).unwrap(), json);
    }

    #[test]
    fn preview_opts_omits_absent_and_round_trips_device() {
        assert_eq!(
            serde_json::to_string(&PreviewOpts::default()).unwrap(),
            "{}"
        );
        let json = r#"{"device":"mobile"}"#;
        let o: PreviewOpts = serde_json::from_str(json).unwrap();
        assert_eq!(o.device, Some(PreviewDevice::Mobile));
        assert_eq!(serde_json::to_string(&o).unwrap(), json);
    }
}

// PORT STATUS: packages/types/src/host/host-bridge.ts (173 lines)
// confidence: medium
// todos: 1
// notes: host bridge contract — not daemon-consumed (low priority per §2.1). Only
// the plain data structs are ported; the behavioral HostBridge/handle interfaces
// are intentionally omitted (they're per-shell runtime code). The TS type-only
// re-exports (Platform/DaemonStatus/LogLevel/UpdateStatus/PresenceState from
// host-contract, DaemonMeta from daemon-target) become `pub use`; AppInfo/Region
// (z.infer aliases) are re-exported from host_contract. Bounds/Region coords → f64
// (CSS px); TerminalOpts cols/rows → i64. Nullable selector/rect/viewport/region
// are required-nullable → Option WITHOUT skip.
